### PR TITLE
Update notes for Spring Security

### PR DESCRIPTION
**Pull request by Excelsior31**
This comprehensive guide on Spring Security covers both REST API and MVC security, highlighting its framework based on Servlet filters for processing web requests and offering declarative and programmatic methods for application security. Key concepts include authentication (verifying user credentials) and authorization (role-based access control). Spring Security is enabled by adding `spring-boot-starter-security` to the pom.xml, which secures all application endpoints and allows for customizations such as CSRF protection (important for traditional web apps but not typically required for stateless REST APIs), database access with a default schema, and BCrypt encryption for secure password handling. For MVC applications, it covers HTTP Basic Authentication, custom login forms using HTML/CSS, logout processes, URL restriction based on roles, and creating a custom access denied page, all integral to securing web applications effectively.

### CODE REVIEW

File: Topics/Tech_Stacks/Spring-Security.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		- username
+		- authority
+
+- BCrypt Encryption
+	- Performs one-way encrypted hashing
+	- Adds a random salt to the password for additional protection
+	- Includes support to defeat brute force attack
+
+	- Generate password
+		- www.luv2code.com/generate-bycrypt-password
```
Comment (XavierQuan) [inline Topics/Tech_Stacks/Spring-Security.md:57]: This link reports 404, please consider removing it or replace it.

File: Topics/Tech_Stacks/Spring-Security.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+- Logout
+	- Logout Process
+		- When a logout is processed, by default Spring Security will
+		- Invalidate user's HTTP session and remove session cookies
+		- Send user back to the login page
+		- Append a logout parameter: ?logout
+
+- Restrict URLs Based on Roles
+	- Dev Process
```
Comment (XavierQuan) [inline Topics/Tech_Stacks/Spring-Security.md:95]: The majority of the content seems to be conceptual. Please consider adding some practical guide like this Dev Process to the above concepts.

**Review by XavierQuan — COMMENTED** [review]
The bulk of the material appears to be theoretical. It would be beneficial to incorporate a practical guide, similar to this Dev Process, alongside the above concepts.

File: Topics/Tech_Stacks/Spring-Security.md (hunk 1 above)
Comment (Excelsior31) [reply Topics/Tech_Stacks/Spring-Security.md:57]: Regarding the broken link, I appreciate you pointing that out. I'll make sure to remove or replace it with a functional and relevant one.

File: Topics/Tech_Stacks/Spring-Security.md (hunk 2 above)
Comment (Excelsior31) [reply Topics/Tech_Stacks/Spring-Security.md:95]: Thank you for your insightful feedback! I absolutely agree that blending practical guides with theoretical content can significantly enhance understanding and application. I'm going to work on including step-by-step processes, much like the 'Dev Process' you mentioned, which should make the material more hands-on and user-friendly.

**Comment by ChenJiang03** [comment]
The current material might be too abstract for new learners who are just beginning to grasp these concepts. Including code snippets would provide concrete examples and make the learning process much more tangible and relatable. This practical approach would not only aid in better comprehension but also enable learners to see how these theoretical concepts are applied in real-world programming scenarios.

**Comment by Excelsior31** [comment]
> The current material might be too abstract for new learners who are just beginning to grasp these concepts. Including code snippets would provide concrete examples and make the learning process much more tangible and relatable. This practical approach would not only aid in better comprehension but also enable learners to see how these theoretical concepts are applied in real-world programming scenarios.


I agree with the comment. Incorporating code snippets into the material is indeed beneficial, especially for beginners, as it transforms abstract concepts into tangible examples. This practical approach enhances understanding and demonstrates the real-world application of theoretical concepts in programming.

File: Topics/Tech_Stacks/Spring-Security.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            .anyRequest().authenticated()
+            .and()
+            .formLogin()
+            .and()
+            .httpBasic();
+    }
+}
+```
+
+- Dev Process
```
Comment (XavierQuan) [inline Topics/Tech_Stacks/Spring-Security.md:48]: The addition of the practical development process greatly enhances the content, making it easier for me to grasp the concept and its application.

**Review by XavierQuan — COMMENTED** [review]
The content has significantly improved following the modifications. Including some graphs would further clarify the concepts and processes, making them more comprehensible.

**Comment by Excelsior31** [comment]
> The content has significantly improved following the modifications. Including some graphs would further clarify the concepts and processes, making them more comprehensible.


I agree with the comment. Adding graphs to the content would indeed enhance clarity and aid in better understanding the concepts and processes. Such visual aids can make complex information more accessible and engaging.

**Review by XavierQuan — APPROVED** [review]
The guide on Spring Security now appears perfect. It seems to be a highly useful resource.

**Comment by Excelsior31** [comment]
> The guide on Spring Security now appears perfect. It seems to be a highly useful resource.

Thank you for your kind words! I'm delighted to hear that the guide meets your expectations and appears to be a valuable resource. If you have any more questions or need further assistance, feel free to ask!

**Review by ChenJiang03 — APPROVED** [review]
Hi! All of the documentation looks great! I am approve you now

**Pull request merged**